plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -94,7 +116,53 @@ private ConnectorSplit toIcebergSplit(FileScanTask task)
                 task.length(),
                 task.file().fileSizeInBytes(),
                 task.file().format(),
-                ImmutableList.of(),
+                getAddresses(task),
                 getPartitionKeys(task));
     }
+
+    private List<HostAddress> getAddresses(FileScanTask task)
+    {
+        if (locality) {
+            return blocklocations(task);
+        }
+        else {
+            return ImmutableList.of();
+        }
+    }
+
+    private List<HostAddress> blocklocations(FileScanTask task)
+    {
+        Set<HostAddress> locationSets = new HashSet<>();
+        Path path = new Path(task.file().path().toString());
+        try {
+            FileSystem fs = hdfsEnvironment.getFileSystem(hdfsContext, path);
+            for (BlockLocation b : fs.getFileBlockLocations(path, task.start(), task.length())) {
+                locationSets.addAll(getHostAddresses(b));
+            }
+        }
+        catch (IOException e) {
+            log.warn("Failed to get block locations for path %s", path, e);
+        }
+        ArrayList<HostAddress> list = new ArrayList<>(locationSets);
```

Comment:
are we assuming that every `FileScanTask` would correspond to always one block? There is a table [property](https://iceberg.apache.org/configuration/#table-properties) defaulting to 128M, but I don't think it's supposed to be aware of HDFS block-size conf. If there're multiple blocks per split, we're returning a randomly ordered list of locations that may correspond to different blocks. So I think it'd be possible that the scheduler may incorrectly assign a split to a worker which doesn't have majority of the split portion locally. 

May be eventually we should "split" the split into multiple splits in that case, like Hive Connector does. To start with, how about just using the first block always? @electrum

Reply:
For ORC (and probably Parquet), we do not sub-split stripes (large physical regions of the ORC file). The split that contains the start of the stripe will read the entire stripe.

As a simple example, suppose the first stripe in a file is 70MB and we produce splits of size 32MB. The first split (for byte range `0..33554432`) will read the entire stripe, since the start of the stripe falls within that range. The second split (for byte range `33554432..67108864`) does not contain the start of any stripe and thus reads nothing (after it reads the ORC file footer to determine the stripe offsets).

A `BlockLocation` has an range (offset and length), so we could pick the blocks that have the largest overlap with the start of the split.

Reply:
However this is (or isn't) handled, there should be a comment explaining why it's done how it is or why it is safe to do that way.

Reply:
Iceberg metadata is supposed to store `split_offset` values for every datafile. This is used during [planning](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/BaseFileScanTask.java#L108) to align iceberg split boundaries with format-specific data boundaries. (e.g. [ORC stripe offsets](https://github.com/apache/iceberg/blob/master/orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java#L108), Parquet row-groups) 

We can implement here keeping in mind this case, where each split is aligned with a stripe/rowgroup. So, we don't need to worry about incorrectly picking an HDFS block corresponding to the "previous" stripe, that the split won't even read. All blocks will always correspond to some part of the split, and hence the stripe. So we should return all blocks that have the highest overlap with the stripe to maximize our chances for locality. (irrespective of whether a block's range contains the "start" of the split) @electrum is this what you meant?

Reply:
Also, note https://github.com/trinodb/trino/issues/9018

Reply:
That makes sense to me.

Reply:
@wenfang6 please look at the above comments.